File: src/test/java/org/veupathdb/lib/blast/BlastConfigContractTest.java
package org.veupathdb.lib.blast;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.veupathdb.lib.blast.consts.Flag;
import org.veupathdb.lib.blast.consts.Key;
import org.veupathdb.lib.blast.field.HSPSorting;
import org.veupathdb.lib.blast.field.HitSorting;
import org.veupathdb.lib.blast.field.OutFormat;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("BlastConfig")
public abstract class BlastConfigContractTest
{
  abstract BlastConfig newConfig();

  @Nested
  @DisplayName("#getTool()")
  class GetTool1
  {
    @Test
    @DisplayName("Does not return null")
    void test1() {
      assertNotNull(newConfig().getTool());
    }
  }

  @Nested
  @DisplayName("-h Flag")
  class ShortHelpFlag
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setShortHelp(null);
      assertNull(test.getShortHelp());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.ShortHelp));

      test.setShortHelp(true);
      assertTrue(test.getShortHelp());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.ShortHelp));
      assertTrue(json.get(Flag.ShortHelp).isBoolean());
      assertTrue(json.get(Flag.ShortHelp).booleanValue());

      test.setShortHelp(false);
      assertFalse(test.getShortHelp());
      json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.ShortHelp));
    }
  }

  @Nested
  @DisplayName("-help Flag")
  class LongHelpFlag
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setLongHelp(null);
      assertNull(test.getLongHelp());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.LongHelp));

      test.setLongHelp(true);
      assertTrue(test.getLongHelp());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.LongHelp));
      assertTrue(json.get(Flag.LongHelp).isBoolean());
      assertTrue(json.get(Flag.LongHelp).booleanValue());

      test.setLongHelp(false);
      assertFalse(test.getLongHelp());
      json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.LongHelp));
    }
  }

  @Nested
  @DisplayName("-version Flag")
  class VersionFlag
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setVersion(null);
      assertNull(test.getVersion());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.Version));

      test.setVersion(true);
      assertTrue(test.getVersion());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.Version));
      assertTrue(json.get(Flag.Version).isBoolean());
      assertTrue(json.get(Flag.Version).booleanValue());

      test.setVersion(false);
      assertFalse(test.getVersion());
      json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.Version));
    }
  }

  @Nested
  @DisplayName("-out Flag")
  class OutFlag
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setOutFile(null);
      assertNull(test.getOutFile());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.OutFile));

      test.setOutFile("Hello");
      assertEquals("Hello", test.getOutFile());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.OutFile));
      assertTrue(json.get(Flag.OutFile).isTextual());
      assertEquals("Hello", json.get(Flag.OutFile).textValue());
    }
  }

  @Nested
  @DisplayName("-outfmt Flag")
  class OutFmt
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setOutFormat(null);
      assertNull(test.getOutFormat());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.OutFormat));

      test.setOutFormat(new OutFormat());
      assertNull(test.getOutFormat());
      json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.OutFormat));

      test.setOutFormat(new OutFormat().setDelimiter("@"));
      assertNotNull(test.getOutFormat());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.OutFormat));
      assertTrue(json.get(Flag.OutFormat).isObject());
      assertEquals(1, json.get(Flag.OutFormat).size());
      assertTrue(json.get(Flag.OutFormat).has(Key.Delimiter));
      assertTrue(json.get(Flag.OutFormat).get(Key.Delimiter).isTextual());
      assertEquals("@", json.get(Flag.OutFormat).get(Key.Delimiter).textValue());
    }
  }

  @Nested
  @DisplayName("-show_gis Flag")
  class ShowGIsFlag
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setShowGIs(null);
      assertNull(test.getShowGIs());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.ShowGIs));

      test.setShowGIs(true);
      assertTrue(test.getShowGIs());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.ShowGIs));
      assertTrue(json.get(Flag.ShowGIs).isBoolean());
      assertTrue(json.get(Flag.ShowGIs).booleanValue());

      test.setShowGIs(false);
      assertFalse(test.getShowGIs());
      json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.ShowGIs));
    }
  }

  @Nested
  @DisplayName("-num_descriptions Flag")
  class NumDescriptions
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setNumDescriptions(null);
      assertNull(test.getNumDescriptions());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.NumDescriptions));

      test.setNumDescriptions(33L);
      assertEquals(33L, test.getNumDescriptions());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.NumDescriptions));
      assertTrue(json.get(Flag.NumDescriptions).isNumber());
      assertEquals(33, json.get(Flag.NumDescriptions).longValue());
    }
  }

  @Nested
  @DisplayName("-num_alignments Flag")
  class NumAlignments
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setNumAlignments(null);
      assertNull(test.getNumAlignments());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.NumAlignments));

      test.setNumAlignments(33L);
      assertEquals(33L, test.getNumAlignments());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.NumAlignments));
      assertTrue(json.get(Flag.NumAlignments).isNumber());
      assertEquals(33, json.get(Flag.NumAlignments).longValue());
    }
  }

  @Nested
  @DisplayName("-line_length Flag")
  class LineLength
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setLineLength(null);
      assertNull(test.getLineLength());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.LineLength));

      test.setLineLength(33);
      assertEquals(33, test.getLineLength());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.LineLength));
      assertTrue(json.get(Flag.LineLength).isNumber());
      assertEquals(33, json.get(Flag.LineLength).intValue());
    }
  }

  @Nested
  @DisplayName("-html Flag")
  class HTMLFlag
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setHTML(null);
      assertNull(test.getHTML());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.HTML));

      test.setHTML(true);
      assertTrue(test.getHTML());
      json = test.toJSON().toJSON();
      assertTrue(json.has(Flag.HTML));
      assertTrue(json.get(Flag.HTML).isBoolean());
      assertTrue(json.get(Flag.HTML).booleanValue());

      test.setHTML(false);
      assertFalse(test.getHTML());
      json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.HTML));
    }
  }

  @Nested
  @DisplayName("-sorthits Flag")
  class SortHits
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setSortHits(null);
      assertNull(test.getSortHits());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.SortHits));

      for (final var hs : HitSorting.values()) {
        test.setSortHits(hs);
        assertEquals(hs, test.getSortHits());
        json = test.toJSON().toJSON();
        assertTrue(json.has(Flag.SortHits));
        assertTrue(json.get(Flag.SortHits).isNumber());
        assertEquals(hs.ordinal(), json.get(Flag.SortHits).intValue());
      }
    }
  }

  @Nested
  @DisplayName("-sorthsps Flag")
  class SortHSPs
  {
    @Test
    @DisplayName("Can be configured")
    void test1() {
      final var test = newConfig();

      test.setSortHSPs(null);
      assertNull(test.getSortHSPs());
      var json = test.toJSON().toJSON();
      assertFalse(json.has(Flag.SortHSPs));

      for (final var hs : HSPSorting.values()) {
        test.setSortHSPs(hs);
        assertEquals(hs, test.getSortHSPs());
        json = test.toJSON().toJSON();
        assertTrue(json.has(Flag.SortHSPs));
        assertTrue(json.get(Flag.SortHSPs).isNumber());
        assertEquals(hs.ordinal(), json.get(Flag.SortHSPs).intValue());
      }
    }
  }

}

File: src/test/java/org/veupathdb/lib/blast/BlastConfigContractTest.kt
package org.veupathdb.lib.blast

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.veupathdb.lib.blast.consts.Flag
import org.veupathdb.lib.blast.consts.Key
import org.veupathdb.lib.blast.field.*
import kotlin.reflect.KClass

@DisplayName("BlastConfig")
abstract class BlastConfigContractTest {
  abstract fun newConfig(): BlastConfig
  abstract fun configClass(): KClass<out BlastConfig>

  @Nested
  @DisplayName("#getTool()")
  internal inner class GetTool1 {
    @Test
    @DisplayName("Does not return null")
    fun test1() {
      Assertions.assertNotNull(newConfig().tool)
    }
  }

  @Nested
  @DisplayName("-h Flag")
  internal inner class ShortHelpFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.shortHelp = null
      Assertions.assertNull(test.shortHelp)
      var json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.ShortHelp))
      test.shortHelp = true
      Assertions.assertTrue(test.shortHelp!!)
      json = test.toJSON().toJSON()
      Assertions.assertTrue(json.has(Flag.ShortHelp))
      Assertions.assertTrue(json[Flag.ShortHelp].isBoolean)
      Assertions.assertTrue(json[Flag.ShortHelp].booleanValue())
      test.shortHelp = false
      Assertions.assertFalse(test.shortHelp!!)
      json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.ShortHelp))
    }
  }

  @Nested
  @DisplayName("-help Flag")
  internal inner class LongHelpFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.longHelp = null
      Assertions.assertNull(test.longHelp)
      var json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.LongHelp))
      test.longHelp = true
      Assertions.assertTrue(test.longHelp!!)
      json = test.toJSON().toJSON()
      Assertions.assertTrue(json.has(Flag.LongHelp))
      Assertions.assertTrue(json[Flag.LongHelp].isBoolean)
      Assertions.assertTrue(json[Flag.LongHelp].booleanValue())
      test.longHelp = false
      Assertions.assertFalse(test.longHelp!!)
      json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.LongHelp))
    }
  }

  @Nested
  @DisplayName("-version Flag")
  internal inner class VersionFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.version = null
      Assertions.assertNull(test.version)
      var json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.Version))
      test.version = true
      Assertions.assertTrue(test.version!!)
      json = test.toJSON().toJSON()
      Assertions.assertTrue(json.has(Flag.Version))
      Assertions.assertTrue(json[Flag.Version].isBoolean)
      Assertions.assertTrue(json[Flag.Version].booleanValue())
      test.version = false
      Assertions.assertFalse(test.version!!)
      json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.Version))
    }
  }

  @Nested
  @DisplayName("-out Flag")
  internal inner class OutFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.outFile = null
      Assertions.assertNull(test.outFile)
      var json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.OutFile))
      test.outFile = OutFile("Hello")
      Assertions.assertEquals("Hello", test.outFile)
      json = test.toJSON().toJSON()
      Assertions.assertTrue(json.has(Flag.OutFile))
      Assertions.assertTrue(json[Flag.OutFile].isTextual)
      Assertions.assertEquals("Hello", json[Flag.OutFile].textValue())
    }
  }

  @Nested
  @DisplayName("-outfmt Flag")
  internal inner class OutFmt {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.outFormat = null
      Assertions.assertNull(test.outFormat)
      var json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.OutFormat))
      test.outFormat = OutFormat()
      Assertions.assertNotNull(test.outFormat)
      json = test.toJSON().toJSON()
      Assertions.assertFalse(json.has(Flag.OutFormat))
      val outFmt = OutFormat()
      outFmt.delimiter = "@"
      test.outFormat = outFmt
      Assertions.assertNotNull(test.outFormat)
      json = test.toJSON().toJSON()
      Assertions.assertTrue(json.has(Flag.OutFormat))
      Assertions.assertTrue(json[Flag.OutFormat].isObject)
      Assertions.assertEquals(1, json[Flag.OutFormat].size())
      Assertions.assertTrue(json[Flag.OutFormat].has(Key.Delimiter))
      Assertions.assertTrue(json[Flag.OutFormat][Key.Delimiter].isTextual)
      Assertions.assertEquals("@", json[Flag.OutFormat][Key.Delimiter].textValue())
    }
  }

  @Nested
  @DisplayName(Flag.ShowGIs)
  internal inner class ShowGIsFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.showGIs = null
      Assertions.assertNull(test.showGIs)
      test.showGIs = true
      Assertions.assertTrue(test.showGIs!!)
      test.showGIs = false
      Assertions.assertFalse(test.showGIs!!)
    }

    @Test
    @DisplayName("Can be deserialized")
    @Throws(
      JsonProcessingException::class)
    fun test2() {
      val input1 = "{\"-show_gis\": true}"
      val test1 = JSON.readValue(input1, configClass().java)!!
      Assertions.assertTrue(test1.showGIs!!)
      val input2 = "{\"-show_gis\": false}"
      val test2 = JSON.readValue(input2, configClass().java)!!
      Assertions.assertFalse(test2.showGIs!!)
    }

    @Test
    @DisplayName("Can be serialized")
    @Throws(
      JsonProcessingException::class)
    fun test3() {
      val test1 = newConfig()
      Assertions.assertEquals("{}", JSON.writeValueAsString(test1))
      test1.showGIs = true
      Assertions.assertEquals("{\"-show_gis\":true}", JSON.writeValueAsString(test1))
      test1.showGIs = false
      Assertions.assertEquals("{}", JSON.writeValueAsString(test1))
    }
  }

  @Nested
  @DisplayName(Flag.NumDescriptions)
  internal inner class NumDescriptionsFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.numDescriptions = null
      Assertions.assertNull(test.numDescriptions)
      test.numDescriptions = NumDescriptions(12L)
      Assertions.assertEquals(12, test.numDescriptions!!.value)
    }

    @Test
    @DisplayName("Can be deserialized")
    @Throws(
      JsonProcessingException::class)
    fun test2() {
      val input = "{\"-num_descriptions\": 32}"
      val parsed = JSON.readValue(input, configClass().java)!!
      Assertions.assertEquals(32, parsed.numDescriptions!!.value)
    }

    @Test
    @DisplayName("Can be serialized")
    @Throws(
      JsonProcessingException::class)
    fun test3() {
      val `object` = newConfig()
      `object`.numDescriptions = NumDescriptions(69L)
      Assertions.assertEquals("{\"-num_descriptions\":69}", JSON.writeValueAsString(`object`))
    }
  }

  @Nested
  @DisplayName(Flag.NumAlignments)
  internal inner class NumAlignmentsFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.numAlignments = null
      Assertions.assertNull(test.numAlignments)
      test.numAlignments = NumAlignments(12L)
      Assertions.assertEquals(12, test.numAlignments!!.value)
    }

    @Test
    @DisplayName("Can be deserialized")
    @Throws(
      JsonProcessingException::class)
    fun test2() {
      val input = "{\"-num_alignments\": 32}"
      val parsed = JSON.readValue(input, configClass().java)!!
      Assertions.assertEquals(32, parsed.numAlignments!!.value)
    }

    @Test
    @DisplayName("Can be serialized")
    @Throws(
      JsonProcessingException::class)
    fun test3() {
      val `object` = newConfig()
      `object`.numAlignments = NumAlignments(69L)
      Assertions.assertEquals("{\"-num_alignments\":69}", JSON.writeValueAsString(`object`))
    }
  }

  @Nested
  @DisplayName(Flag.LineLength)
  internal inner class LineLengthFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.lineLength = null
      Assertions.assertNull(test.lineLength)
      test.lineLength = LineLength(12)
      Assertions.assertEquals(12, test.lineLength!!.value)
    }

    @Test
    @DisplayName("Can be deserialized")
    @Throws(
      JsonProcessingException::class)
    fun test2() {
      val input = "{\"-line_length\": 32}"
      val parsed = JSON.readValue(input, configClass().java)!!
      Assertions.assertEquals(32, parsed.lineLength!!.value)
    }

    @Test
    @DisplayName("Can be serialized")
    @Throws(
      JsonProcessingException::class)
    fun test3() {
      val `object` = newConfig()
      `object`.lineLength = LineLength(69)
      Assertions.assertEquals("{\"-line_length\":69}", JSON.writeValueAsString(`object`))
    }
  }

  @Nested
  @DisplayName(Flag.HTML)
  internal inner class HTMLFlag {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.html = null
      Assertions.assertNull(test.html)
      test.html = true
      Assertions.assertTrue(test.html!!)
      test.html = false
      Assertions.assertFalse(test.html!!)
    }

    @Test
    @DisplayName("Can be deserialized")
    @Throws(
      JsonProcessingException::class)
    fun test2() {
      val input1 = "{\"-html\": true}"
      val test1 = JSON.readValue(input1, configClass().java)!!
      Assertions.assertTrue(test1.html!!)
      val input2 = "{\"-html\": false}"
      val test2 = JSON.readValue(input2, configClass().java)!!
      Assertions.assertFalse(test2.html!!)
    }

    @Test
    @DisplayName("Can be serialized")
    @Throws(
      JsonProcessingException::class)
    fun test3() {
      val test1 = newConfig()
      Assertions.assertEquals("{}", JSON.writeValueAsString(test1))
      test1.html = true
      Assertions.assertEquals("{\"-html\":true}", JSON.writeValueAsString(test1))
      test1.html = false
      Assertions.assertEquals("{}", JSON.writeValueAsString(test1))
    }
  }

  @Nested
  @DisplayName(Flag.SortHits)
  internal inner class SortHits {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.sortHits = null
      Assertions.assertNull(test.sortHits)
      for (hs in HitSorting.values()) {
        test.sortHits = hs
        Assertions.assertEquals(hs, test.sortHits)
      }
    }

    @Test
    @DisplayName("Can be serialized")
    @Throws(
      JsonProcessingException::class)
    fun test2() {
      val test = newConfig()
      for (`val` in HitSorting.values()) {
        test.sortHits = `val`
        Assertions.assertEquals(
          "{\"" + Flag.SortHits + "\":" + `val`.ordinal + "}",
          JSON.writeValueAsString(test)
        )
      }
    }

    @Test
    @DisplayName("Can be deserialized")
    @Throws(
      JsonProcessingException::class)
    fun test3() {
      for (`val` in HitSorting.values()) {
        val json = "{\"" + Flag.SortHits + "\":" + `val`.ordinal + "}"
        val test = JSON.readValue(json, configClass().java)!!
        Assertions.assertEquals(`val`, test.sortHits)
      }
    }
  }

  @Nested
  @DisplayName(Flag.SortHSPs)
  internal inner class SortHSPs {
    @Test
    @DisplayName("Can be configured")
    fun test1() {
      val test = newConfig()
      test.sortHSPs = null
      Assertions.assertNull(test.sortHSPs)
      for (hs in HSPSorting.values()) {
        test.sortHSPs = hs
        Assertions.assertEquals(hs, test.sortHSPs)
      }
    }

    @Test
    @DisplayName("Can be serialized")
    @Throws(
      JsonProcessingException::class)
    fun test2() {
      val test = newConfig()
      for (`val` in HSPSorting.values()) {
        test.sortHSPs = `val`
        Assertions.assertEquals(
          "{\"" + Flag.SortHSPs + "\":" + `val`.ordinal + "}",
          JSON.writeValueAsString(test)
        )
      }
    }

    @Test
    @DisplayName("Can be deserialized")
    @Throws(
      JsonProcessingException::class)
    fun test3() {
      for (`val` in HSPSorting.values()) {
        val json = "{\"" + Flag.SortHSPs + "\":" + `val`.ordinal + "}"
        val test = JSON.readValue(json, configClass().java)!!
        Assertions.assertEquals(`val`, test.sortHSPs)
      }
    }
  }

  companion object {
    private val JSON = ObjectMapper()
  }
}